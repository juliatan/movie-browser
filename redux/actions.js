import { fetchMovies } from '../api';

// action types -> done to prevent typos and for easier bug identification
export const SEARCH_TERM_SENT = 'SEARCH_TERM_SENT'
export const NEW_RESULTS_RECEIVED = 'NEW_RESULTS_RECEIVED'
export const ADDITIONAL_RESULTS_RECEIVED = 'ADDITIONAL_RESULTS_RECEIVED'
export const RESULTS_ERROR = 'RESULTS_ERROR'
export const CLEAR_RESULTS = 'CLEAR_RESULTS'

// async action creator

// invoked directly from SearchScreen.js. This dispatches various different "types" to reducer.js.
// Because dispatch is a function, we need to use 3rd party "redux-thunk" library and import "applyMiddleware" from redux in store.js.
export const returnResults = (searchTerm, pageNumber) => async dispatch => {
  dispatch({type: SEARCH_TERM_SENT})
  try {
    const response = await fetchMovies(searchTerm, pageNumber)
    const results = response.results

    const totalResults = response.totalResults
    const maxPages = Math.ceil(totalResults/10) // 10 results per page

    if (pageNumber === 1) {
      dispatch({type: NEW_RESULTS_RECEIVED, payload: { movies: results, maxPages: maxPages }})
    } else {
      dispatch({type: ADDITIONAL_RESULTS_RECEIVED, payload: { movies: results }})
    }

  } catch (err) {
    console.error('API error')
    dispatch({type: RESULTS_ERROR, payload: 'API error'})
  }
}

export const clearResults = () => dispatch => {
  dispatch({type: CLEAR_RESULTS})
}