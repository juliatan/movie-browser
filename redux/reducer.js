import { combineReducers } from 'redux'
import { SEARCH_TERM_SENT, NEW_RESULTS_RECEIVED, ADDITIONAL_RESULTS_RECEIVED, RESULTS_ERROR } from './actions'

const merge = (prev, next) => Object.assign({}, prev, next)

// This is listening for various dispatch "types" of actions. Depending on type, the reducer updates the state, so it can
// be accessed back in SearchScreen.js via the mapStateToProps function.
const resultsReducer = (state = {}, action) => {
  switch (action.type) {
    case NEW_RESULTS_RECEIVED:
      return merge(state, { movies: action.payload.movies, maxPages: action.payload.maxPages })
    case ADDITIONAL_RESULTS_RECEIVED:
      return {...state, movies: [...state.movies, ...action.payload.movies]}
    case RESULTS_ERROR:
      return merge(state, { resultsErr: action.payload })
    default:
      return state
  }
}

// "results" key is what mapStateToProps function reads as "state.results"
const reducer = combineReducers({
  results: resultsReducer
})

export default reducer