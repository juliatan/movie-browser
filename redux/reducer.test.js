import reducer from './reducer'
import * as actions from './actions'

const DEFAULT_STATE = {
  movies: null,
  maxPages: null
}

describe ('results reducer', () => {
  // note run for the first time to save snapshot that takes the form you want it to be
  it('successfully receives new results', () => {
    expect(reducer(DEFAULT_STATE, actions.returnResults({
      movies: [],
      maxPages: 2
    }))).toMatchSnapshot()
  })
})