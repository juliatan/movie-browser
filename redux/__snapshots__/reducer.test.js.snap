// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`results reducer successfully receives new results 1`] = `
Object {
  "results": Object {},
}
`;
